File: sdks/py/src/jeffs_brain_memory/memory/slugmap.py
# SPDX-License-Identifier: Apache-2.0
"""Machine-local map of project path -> canonical slug."""

from __future__ import annotations

import os
import threading
from pathlib import Path

import yaml


def _default_path() -> str:
    home = os.path.expanduser("~")
    if not home or home == "~":
        home = "."
    return str(Path(home) / ".local" / "state" / "jeffs-brain" / "slug-map.yaml")


class SlugMap:
    """Persistent path -> slug mapping."""

    def __init__(self, path: str = "") -> None:
        self._lock = threading.Lock()
        self._entries: dict[str, str] = {}
        self._path = path or _default_path()

    @property
    def path(self) -> str:
        return self._path

    def load(self) -> None:
        with self._lock:
            try:
                raw = Path(self._path).read_bytes()
            except OSError:
                return
            try:
                parsed = yaml.safe_load(raw.decode("utf-8")) or {}
            except yaml.YAMLError:
                return
            if isinstance(parsed, dict):
                self._entries = {str(k): str(v) for k, v in parsed.items()}

    def save(self) -> None:
        with self._lock:
            self._save_locked()

    def _save_locked(self) -> None:
        target = Path(self._path)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(yaml.safe_dump(self._entries, sort_keys=True), encoding="utf-8")

    def lookup(self, project_path: str) -> str | None:
        with self._lock:
            return self._entries.get(project_path)

    def register(self, project_path: str, slug: str) -> None:
        with self._lock:
            self._entries[project_path] = slug
            self._save_locked()

File: go/memory/slugmap.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"os"
	"path/filepath"
	"sync"

	"gopkg.in/yaml.v3"
)

// SlugMap maps local project paths to canonical slugs. Each machine keeps
// its own map because filesystem paths are inherently machine-specific;
// only the resulting canonical slug travels with the brain store.
type SlugMap struct {
	mu      sync.Mutex
	entries map[string]string
	path    string
}

// slugMapPath returns the default location for the machine-local slug
// map: ~/.local/state/jeffs-brain/slug-map.yaml.
func slugMapPath() string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return filepath.Join(home, ".local", "state", "jeffs-brain", "slug-map.yaml")
}

// NewSlugMap creates a SlugMap backed by the given file path. Pass an
// empty string to use the default location.
func NewSlugMap(path string) *SlugMap {
	if path == "" {
		path = slugMapPath()
	}
	return &SlugMap{
		entries: make(map[string]string),
		path:    path,
	}
}

// Load reads the slug map from disk. Missing or unreadable files are
// silently ignored.
func (m *SlugMap) Load() {
	m.mu.Lock()
	defer m.mu.Unlock()

	data, err := os.ReadFile(m.path)
	if err != nil {
		return
	}

	parsed := make(map[string]string)
	if err := yaml.Unmarshal(data, &parsed); err != nil {
		return
	}
	m.entries = parsed
}

// Save writes the slug map to disk, creating parent directories as
// needed.
func (m *SlugMap) Save() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.saveLocked()
}

func (m *SlugMap) saveLocked() error {
	if err := os.MkdirAll(filepath.Dir(m.path), 0o755); err != nil {
		return err
	}

	data, err := yaml.Marshal(m.entries)
	if err != nil {
		return err
	}
	return os.WriteFile(m.path, data, 0o644)
}

// Lookup returns the canonical slug for the given project path.
func (m *SlugMap) Lookup(projectPath string) (string, bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	slug, ok := m.entries[projectPath]
	return slug, ok
}

// Register stores a path to slug mapping and persists the map.
func (m *SlugMap) Register(projectPath, slug string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.entries[projectPath] = slug
	return m.saveLocked()
}
